File: experiment-14/runs/language=elixir_model=sonnet_prompt=TDD/rep1/test/brazilian_soccer/queries/matches_test.exs
defmodule BrazilianSoccer.Queries.MatchesTest do
  use ExUnit.Case, async: false

  alias BrazilianSoccer.Queries.Matches

  describe "search_by_team/1" do
    test "finds matches for a team as home or away" do
      results = Matches.search_by_team("Flamengo")
      assert length(results) > 0
      assert Enum.all?(results, fn m ->
        String.contains?(m.home_team, "Flamengo") or String.contains?(m.away_team, "Flamengo")
      end)
    end

    test "is case-insensitive" do
      results = Matches.search_by_team("flamengo")
      assert length(results) > 0
    end

    test "returns empty list for unknown team" do
      results = Matches.search_by_team("NonExistentTeamXYZ")
      assert results == []
    end
  end

  describe "search_by_teams/2" do
    test "finds head-to-head matches between two teams" do
      results = Matches.search_by_teams("Flamengo", "Fluminense")
      assert length(results) > 0
      assert Enum.all?(results, fn m ->
        (String.contains?(m.home_team, "Flamengo") and String.contains?(m.away_team, "Fluminense")) or
        (String.contains?(m.home_team, "Fluminense") and String.contains?(m.away_team, "Flamengo"))
      end)
    end
  end

  describe "search_by_competition/1" do
    test "finds matches in Brasileirão" do
      results = Matches.search_by_competition("Brasileirão")
      assert length(results) > 0
      assert Enum.all?(results, fn m -> m.competition == "Brasileirão" end)
    end

    test "finds matches in Copa do Brasil" do
      results = Matches.search_by_competition("Copa do Brasil")
      assert length(results) > 0
    end

    test "finds matches in Copa Libertadores" do
      results = Matches.search_by_competition("Copa Libertadores")
      assert length(results) > 0
    end

    test "case-insensitive partial match" do
      results = Matches.search_by_competition("libertadores")
      assert length(results) > 0
    end
  end

  describe "search_by_season/1" do
    test "finds matches in a specific season" do
      results = Matches.search_by_season(2019)
      assert length(results) > 0
      assert Enum.all?(results, fn m -> m.season == 2019 end)
    end

    test "returns empty for season with no data" do
      results = Matches.search_by_season(1800)
      assert results == []
    end
  end

  describe "search_by_team_and_season/2" do
    test "finds Palmeiras matches in 2023" do
      results = Matches.search_by_team_and_season("Palmeiras", 2023)
      assert length(results) > 0
      assert Enum.all?(results, fn m ->
        (String.contains?(m.home_team, "Palmeiras") or String.contains?(m.away_team, "Palmeiras")) and
        m.season == 2023
      end)
    end
  end

  describe "head_to_head_stats/2" do
    test "returns win/draw/loss record for two teams" do
      stats = Matches.head_to_head_stats("Flamengo", "Fluminense")
      assert is_map(stats)
      assert Map.has_key?(stats, :flamengo_wins)
      assert Map.has_key?(stats, :fluminense_wins)
      assert Map.has_key?(stats, :draws)
      assert Map.has_key?(stats, :total)
      assert stats.total == stats.flamengo_wins + stats.fluminense_wins + stats.draws
    end
  end

  describe "biggest_wins/2" do
    test "returns matches sorted by goal difference" do
      results = Matches.biggest_wins(5)
      assert length(results) == 5
      diffs = Enum.map(results, fn m -> abs(m.home_goal - m.away_goal) end)
      assert diffs == Enum.sort(diffs, :desc)
    end
  end

  describe "search_by_date_range/2" do
    test "finds matches within a date range" do
      results = Matches.search_by_date_range("2023-01-01", "2023-12-31")
      assert length(results) > 0
    end

    test "returns empty for out-of-range dates" do
      results = Matches.search_by_date_range("1800-01-01", "1800-12-31")
      assert results == []
    end
  end
end

File: experiments/adrianco/experiment-5/runs/language=typescript_model=claude-opus-4-7_tooling=none/rep3/tests/matches.test.ts
import { describe, it, expect, beforeAll } from 'vitest';
import { loadStore } from './setup.js';
import { findMatches, headToHead, biggestWins } from '../src/queries/matches.js';
import type { DataStore } from '../src/types.js';

/**
 * Feature: Match Queries
 *
 * Scenario: Find matches between two teams
 *   Given the match data is loaded
 *   When I search for matches between "Flamengo" and "Fluminense"
 *   Then I should receive a list of matches
 *   And each match should have date, scores, and competition
 */
describe('Feature: Match Queries', () => {
  let store: DataStore;
  beforeAll(async () => {
    store = await loadStore();
  });

  describe('Scenario: Find matches between two teams', () => {
    it('returns Flamengo vs Fluminense matches with dates, scores, competition', () => {
      const ms = findMatches(store.matches, {
        team: 'Flamengo',
        opponent: 'Fluminense',
      });
      expect(ms.length).toBeGreaterThan(5);
      for (const m of ms) {
        expect(m.date).toMatch(/^\d{4}-\d{2}-\d{2}$/);
        expect(typeof m.homeGoals).toBe('number');
        expect(typeof m.awayGoals).toBe('number');
        expect(['Brasileirão', 'Copa do Brasil', 'Libertadores', 'Other']).toContain(
          m.competition,
        );
        // Both teams must be involved
        const both = [m.homeKey, m.awayKey];
        expect(both).toContain('flamengo');
        expect(both).toContain('fluminense');
      }
    });
  });

  describe('Scenario: Find matches by team and season', () => {
    it('finds Palmeiras matches in 2019', () => {
      const ms = findMatches(store.matches, {
        team: 'Palmeiras',
        season: 2019,
      });
      expect(ms.length).toBeGreaterThan(10);
      for (const m of ms) {
        expect(m.season).toBe(2019);
        expect([m.homeKey, m.awayKey]).toContain('palmeiras');
      }
    });
  });

  describe('Scenario: Find matches by competition', () => {
    it('returns only Libertadores matches when filter set', () => {
      const ms = findMatches(store.matches, {
        competition: 'Libertadores',
        limit: 50,
      });
      expect(ms.length).toBeGreaterThan(0);
      for (const m of ms) {
        expect(m.competition).toBe('Libertadores');
      }
    });
  });

  describe('Scenario: Filter to home-only matches', () => {
    it('returns only home matches when homeOnly=true', () => {
      const ms = findMatches(store.matches, {
        team: 'Corinthians',
        season: 2018,
        competition: 'Brasileirão',
        homeOnly: true,
      });
      expect(ms.length).toBeGreaterThan(0);
      for (const m of ms) {
        expect(m.homeKey).toContain('corinthians');
      }
    });
  });

  describe('Scenario: Date range filtering', () => {
    it('respects dateFrom and dateTo bounds', () => {
      const ms = findMatches(store.matches, {
        dateFrom: '2019-01-01',
        dateTo: '2019-12-31',
        limit: 100,
      });
      expect(ms.length).toBeGreaterThan(0);
      for (const m of ms) {
        expect(m.date >= '2019-01-01').toBe(true);
        expect(m.date <= '2019-12-31').toBe(true);
      }
    });
  });

  describe('Scenario: Head-to-head record between two teams', () => {
    it('counts wins, draws, and aggregates goals', () => {
      const h = headToHead(store.matches, 'Flamengo', 'Fluminense');
      expect(h.matches.length).toBeGreaterThan(0);
      expect(h.teamAWins + h.teamBWins + h.draws).toBe(h.matches.length);
      expect(h.teamAGoals).toBeGreaterThanOrEqual(0);
      expect(h.teamBGoals).toBeGreaterThanOrEqual(0);
    });
  });

  describe('Scenario: Biggest victories in the dataset', () => {
    it('returns matches ordered by goal margin', () => {
      const ms = biggestWins(store.matches, { limit: 10 });
      expect(ms.length).toBe(10);
      const margins = ms.map((m) => Math.abs(m.homeGoals - m.awayGoals));
      for (let i = 1; i < margins.length; i++) {
        expect(margins[i] <= margins[i - 1]).toBe(true);
      }
      // The biggest margin in this dataset is at least 5 goals.
      expect(margins[0]).toBeGreaterThanOrEqual(5);
    });
  });
});
